### PR TITLE
Be able to customize default widgets

### PR DESCRIPTION
**Pull request by frassinier**
### Reasons for making this change

Provide a registry service in order to let customizing default widgets (or whatever).
It needs some polish but it works.

That pull request is related to #358 and #363.
And because I don't want to deal with all logic by manipulating fields like with #283.  
### Demo

Let's try to implement `material-ui` to change our project playground aspect,
(Start by editing _app.js_)

``` javascript
// Example of default Checkbox widget override
import WidgetRegistry from "../src/registry/WidgetRegistry";
import MuiThemeProvider from "material-ui/styles/MuiThemeProvider";
import MaterialDesignCheckboxWidget from "./themes/material-design/CheckboxWidget";

WidgetRegistry.addToDefaultWidgetMap("boolean", MaterialDesignCheckboxWidget);
```

Redefine pure Checkbox widget (_themes/material-design/CheckboxWidget.js_) 

``` javascript
import React from "react";
import Checkbox from "material-ui/Checkbox";

function MDCheckboxWidget({
    id,
    value,
    disabled,
    label,
    onChange,
}) {
    return (
        <Checkbox
            id={id}
            checked={!!value}
            disabled={disabled}
            onCheck={(event) => onChange(event.target.checked)}
            label={label}
        />
    );
}

export default MDCheckboxWidget;
```

Because `material-ui`, we need to wrap with MuiThemeProvider (_app.js_)

``` javascript
<MuiThemeProvider>
    <Form schema={schema}
        uiSchema={uiSchema}
        formData={formData}/>
</MuiThemeProvider>}
```

That's all folks! :)

checkbox md

Could you imagine how it will be easy to create themes?
### Checklist
- [ ] **I'm updating documentation**
  - [ ] I've checked the rendering of the Markdown text I've added
  - [ ] If I'm adding a new section, I've updated the Table of Content
- [x] **I'm adding or updating code**
  - [ ] I've added and/or updated tests
  - [ ] I've updated docs if needed
- [x] **I'm adding a new feature**
  - [x] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by jmfrancois** [comment]
The main idea is to keep the lib as it is and let anyone who want to change all default widgets to do so without forking the lib.

We have tried different ways and this one is the best because it let the lib handle the behavior and just change the defaults;

It has a real benefit for us (we want material design in our forms).
Thanks @frassinier for this great job.

PS: hello @leplatrem as an old colleague from makina ;)

**Comment by olzraiti** [comment]
Related also closely to #283, #310 and #354.

File: src/registry/Registry.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+/**
+ * The registry that will have the singleton
+ * - getRegistry() -> the registry which is a simple key/value POJO
+ * @type {Object}
+ */
+const Registry = {
```
Comment (n1k0) [inline src/registry/Registry.js]: So if this becomes a singleton, how about using multiple forms using different registries in a same page?

Comment (frassinier) [inline src/registry/Registry.js]: It's a widget registry, just a hook to redefine all pure components (TextWidget, CheckboxWidget, etc.). 
I can't imagine that you would start using `react-bootstrap` for one form and `material-ui` for another one :/

**Comment by frassinier** [comment]
> I don't like the global registry singleton approach; could we imagine having a single one per form? eg. reusing shouldn't be an issue as we could pass a constructed one to multiple form instances

No, I think it's library related. "As a developer, I would set `material-ui` theme for all forms in my app" 

> This adds a lot of code, getter and setters, which hasn't been the general style for this project; I understand that's the style you prefer, but could we conceived trying to adapt it a little more to the existing one?

If this approach is kept, I will simplify the code.

**Comment by n1k0** [comment]
> No, I think it's library related. "As a developer, I would set material-ui theme for all forms in my app"

I disagree; how about we let developers deciding about that? Instantiating a registry object instead of relying on some stateful global sounds like a much saner approach to me, while being as powerful (also, just think about testing: resetting the singleton state before each test, really?).

**Comment by frassinier** [comment]
Ok, maybe it's only my need, so what's about adding a prop `theme` to Form which can edit the default widget mapping from _utils.js_?

**Comment by n1k0** [comment]
That sounds like a very convenient API indeed :)

**Comment by jmfrancois** [comment]
@n1k0 I think this is a different scope.

* be able to change the defaults widgets
* be able to change the default widgets per form

AFAIK you already can do a change for one form by using the schema field.

This is a really common need to set the look & feel of all forms in your app.
I would find this really boring for each form to add *theme* just to change the default.

The registry with getter and setter has nothing related with code style.
Doing import of a widget in a field to say this is the default widget for that field is hardcoded default. I don't know any other way to let integrator changing defaults except by a registry.

**Comment by n1k0** [comment]
```js
const registry = new Registry({...));
render(<div><Form registry={registry} /><Form registry={registry} /></div>)
```

vs

```js
render(<div><Form /><Form /></div>)
```

I'm sorry, are we really arguing over this? Please excuse my bluntness, but I don't want global singletons, period.

**Comment by frassinier** [comment]
You definitely right, even if I propose to set `theme` rather than `registry` prop.

Your third snippet is what we will do as high order function to fit to our needs and I understand it's not what all users want.

I plan to edit widget mapping from _utils.js_ and if a `theme` prop i provided then lib will merge mappings. Looks good to you?

**Comment by jmfrancois** [comment]
> I'm sorry, are we really arguing over this?

No we just try to understand your review.
I had not seen the fact that a registry imply a kind of state, which is bad it's true.
Our goal was to change the default so having an API to do that.

So let's move from a global registry for default to simple props.

**Comment by n1k0** [comment]
> I plan to edit widget mapping from utils.js and if a theme prop i provided then lib will merge mappings. Looks good to you?

Most definitely! Thanks

**Comment by frassinier** [comment]
It's up to date @n1k0

**Comment by n1k0** [comment]
Thanks, I'll give this a proper review tomorrow.

**Comment by frassinier** [comment]
[Here](https://github.com/frassinier/rjsf-material-design) is an example of a possible theme.

```javascript
import Form from "react-jsonschema-form";

import customizedWidgets from "rjsf-material-design";

<Form schema={schema}
      uiSchema={uiSchema}
      formData={formData}
      theme={customDefaultWidgets}/>
```

File: src/components/fields/SchemaField.js [hunk 2]
```diff
@@ -147,6 +148,10 @@ function SchemaField(props) {
     return <div/>;
   }
 
+  if (theme) {
+    mergeToDefaultWidgets(theme);
```
Comment (n1k0) [inline src/components/fields/SchemaField.js]: That's a massive side effect. I'd rather at least have a new object retrieved from the call to this function:

```js
const themedWidgets = mergeToDefaultWidgets(theme);
```

File: src/utils.js [hunk 3]
```diff
@@ -93,6 +113,36 @@ export function defaultFieldValue(formData, schema) {
...
+  };
+  return widget;
+}
+
+export function getDefaultWidget(
+  schema,
+  widgetType,
+) {
+  const {type} = schema;
+  if(!widgetType && Array.isArray(schema.enum)) {
```
Comment (n1k0) [inline src/utils.js]: Nit: space after `if`

**Comment by frassinier** [comment]
To move forward, that feature will be added with #371

**Pull request closed without merging**